src/framed_write.rs

```diff
@@ -109,11 +115,26 @@ where
     }
     fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<(), Self::Error>> {
         let this = &mut *self;
-        ready!(Pin::new(&mut this.inner).poll_write(cx, &this.buffer))?;
-        Pin::new(&mut self.inner).poll_flush(cx).map_err(Into::into)
+
+        ready!(Pin::new(&mut this.inner).poll_flush(cx).map_err(Into::into))?;
+
+        while !this.buffer.is_empty() {
+            let num_write = ready!(Pin::new(&mut this.inner).poll_write(cx, &this.buffer))?;
+
+            if num_write == 0 {
+                return Poll::Ready(Err(
+                    Error::new(ErrorKind::UnexpectedEof, "End of file").into()
+                ));
+            }
+
+            let _ = this.buffer.split_to(num_write);
+            ready!(Pin::new(&mut this.inner).poll_flush(cx).map_err(Into::into))?;
+        }
+        Poll::Ready(Ok(()))
     }
     fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<(), Self::Error>> {
-        Pin::new(&mut self.inner).poll_close(cx).map_err(Into::into)
+        ready!(Pin::new(&mut *self).poll_flush(cx))?;
+        Pin::new(&mut (*self).inner).poll_close(cx).map_err(Into::into)
```

Comment:
This is good, but I would write it like this to get rid of all the de-referencing:
```rust
ready!(self.as_mut().poll_flush(cx))?;
Pin::new(&mut self.inner).poll_close(cx).map_err(Into::into)
```

Reply:
Done in ffc4687f20e15db939e574a53d49998e093e2bad.